### PR TITLE
Fixed some comparison issues

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed
Improved comparison which gives better error messages when using symbolic variables where it is not meant to be.

#### Other comments
There are lots of comparisons, so these are only a few. But better do it carefully in small batches...

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/calculus/finite_diff.py [hunk 1]
```diff
@@ -167,6 +167,9 @@ def finite_diff_weights(order, x_list, x0=S.One):
 
     """
     # The notation below closely corresponds to the one used in the paper.
+    order = S(order)
+    if not order.is_number:
+        raise ValueError("Cannot handle symbolic order.")
```
Comment (oscarbenjamin) [inline sympy/calculus/finite_diff.py]: Can these three conditions be combined? After simpification the result should be a non-negative Integer so:
```python
if not order.is_Integer and order >= 0:
     raise ValueError("Order must be an explicit nonnegative integer.")
```

File: sympy/concrete/tests/test_products.py [hunk 2]
```diff
@@ -234,6 +234,12 @@ def test__eval_product():
     assert product(2*a(i), (i, 1, n)) == 2**n * Product(a(i), (i, 1, n))
     # issue 4810
     assert product(2**i, (i, 1, n)) == 2**(n/2 + n**2/2)
+    k, m = symbols('k m', integer=True)
+    assert product(2**i, (i, k, m)) == 2**(-k**2/2 + k/2 + m**2/2 + m/2)
+    n = Symbol('n', negative=True, integer=True)
+    p = Symbol('p', positive=True, integer=True)
+    assert product(2**i, (i, n, p)) == 2**(-n**2/2 + n/2 + p**2/2 + p/2)
+    assert product(2**i, (i, p, n)) == 2**(n**2/2 + n/2 - p**2/2 + p/2)
```
Comment (oscarbenjamin) [inline sympy/concrete/tests/test_products.py]: How is the product defined when the end index is less than the start? Is it just the reciprocal of the product with limits flipped? Does something similar happen for sums?

File: sympy/algebras/quaternion.py [hunk 3]
```diff
@@ -421,22 +420,23 @@ def pow(self, p):
...
         668 + (-224)*i + (-336)*j + (-448)*k
         """
+        p = sympify(p)
         q = self
         if p == -1:
             return q.inverse()
         res = 1
 
+        if fuzzy_not(p.is_Integer):
+            return NotImplemented
```
Comment (oscarbenjamin) [inline sympy/algebras/quaternion.py]: I think `p.is_Integer` is equivalent to `isinstance(p, Integer)` so is always (non-symbolic) True or False. This should be generally true of any `is_Capitalised` attribute but this is not always done consistently. I don't think that the `fuzzy_not` is needed here.

Comment (smichr) [reply sympy/algebras/quaternion.py]: Yes. is_integer could be None but not is_Integer

Comment (oscargus) [reply sympy/algebras/quaternion.py]: OK, will fix!

File: sympy/concrete/tests/test_products.py (hunk 2 above)
Comment (oscargus) [reply sympy/concrete/tests/test_products.py]: Products:
https://github.com/sympy/sympy/blob/a387ec651951f72b77ca45b3a43c0a45bc142dfe/sympy/concrete/products.py#L210-L214

Sums:
https://github.com/sympy/sympy/blob/a387ec651951f72b77ca45b3a43c0a45bc142dfe/sympy/concrete/summations.py#L190-L194

I do not fully understand it to be honest. The change here (motivating the test), was that it currently only swapped with numerical limits. Not sure that the current test actually tests it, but at least it should exercise the code in a new way (it is the same result as earlier).

File: sympy/calculus/finite_diff.py (hunk 1 above)
Comment (oscargus) [reply sympy/calculus/finite_diff.py]: Yes, for trading clarity in error message.

**Comment by oscarbenjamin** [comment]
Looks good. Merging

**Pull request merged**